lib/compress/zstd_double_fast.c

```diff
@@ -312,8 +359,12 @@ size_t ZSTD_compressBlock_doubleFast_dictMatchState_generic(
         U32 offset;
         size_t const h2 = ZSTD_hashPtr(ip, hBitsL, 8);
         size_t const h = ZSTD_hashPtr(ip, hBitsS, mls);
-        size_t const dictHL = ZSTD_hashPtr(ip, dictHBitsL, 8);
-        size_t const dictHS = ZSTD_hashPtr(ip, dictHBitsS, mls);
+        size_t const dictHashAndTagL = ZSTD_hashPtr(ip, dictHBitsL + ZSTD_SHORT_CACHE_TAG_BITS, 8);
+        size_t const dictHashAndTagS = ZSTD_hashPtr(ip, dictHBitsS + ZSTD_SHORT_CACHE_TAG_BITS, mls);
+        U32 const dictMatchIndexAndTagL = dictHashLong[dictHashAndTagL >> ZSTD_SHORT_CACHE_TAG_BITS];
+        U32 const dictMatchIndexAndTagS = dictHashSmall[dictHashAndTagS >> ZSTD_SHORT_CACHE_TAG_BITS];
```

Comment:
I see you're now unconditionally doing these reads where before they were only issued if they working ctx search failed. I guess that fits with the pipelining theme... But yeah I'm slightly curious how you arrived at this choice.

Reply:
I think this fits with the general pattern we've talked about of hoisting memory above logic. When the reads are inside branches, the compiler can't safely move them up.

Also, because DMS runs on small inputs, the ctx search will fail most of the time, so most of the time the reads will happen anyway. And because doublefast DMS is not pipelined, there is likely load bandwidth sitting unused, so it's probably not too expensive anyway.

However, the answer ultimately comes down to "I tried this and it's fast" :)

Reply:
Makes sense.